pkg/storaged/crypto-keyslots.jsx

```diff
@@ -0,0 +1,556 @@
+                        </label>
+                    </div>
+                    <div className="radio-child" hidden={val !== false}>
+                        Removing a passphrase without confirmation can be dangerous.
+                    </div>
+                </div>
+            );
+        }
+    };
+};
+
+function remove_passphrase_dialog(block, key) {
+    dialog_open({ Title: _("Remove passphrase"),
+                  Fields: [
+                      RemovePassphraseField("passphrase", key, block_name(block))
+                  ],
+                  Action: {
+                      DangerButton: true,
+                      Title: _("Remove"),
+                      action: function (vals) {
+                          if (vals.passphrase === false)
+                              return slot_remove(block, key.slot);
+                          else
+                              return passphrase_remove(block, vals.passphrase);
+                      }
+                  }
+    });
+}
+
+const RemoveClevisField = (tag, key, dev) => {
+    return {
+        tag: tag,
+        title: <span className="pficon pficon-warning-triangle-o large" />,
+        options: { },
+        initial_value: "",
+
+        render: (val, change) => {
+            return (
+                <div data-field={tag}>
+                    <div className="large">{ fmt_to_array(_("Remove $0?"), <b>{key.url}</b>) }</div>
+                    <div>{ fmt_to_array(_("Keyserver removal may prevent unlocking $0."), <b>{dev}</b>) }</div>
+                </div>
+            );
+        }
+    };
+};
+
+function remove_clevis_dialog(client, block, key) {
+    dialog_open({ Title: _("Remove Tang keyserver"),
+                  Fields: [
+                      RemoveClevisField("keyserver", key, block_name(block))
+                  ],
+                  Action: {
+                      DangerButton: true,
+                      Title: _("Remove"),
+                      action: function () {
+                          return clevis_remove(block, key);
+                      }
+                  }
+    });
+}
+
+export class CryptoKeyslots extends React.Component {
+    constructor() {
+        super();
+        this.state = { version: 1, slots: null, slot_error: null, max_slots: 8 };
+    }
+
+    monitor_slots(block) {
+        // HACK - we only need this until UDisks2 has a Encrypted.Slots property or similar.
+        if (block != this.monitored_block) {
+            if (this.monitored_block)
+                this.monitor_channel.close();
+            this.monitored_block = block;
+            if (block) {
+                var dev = decode_filename(block.Device);
+                this.monitor_channel = python.spawn(luksmeta_monitor_hack_py, [ dev ],
+                                                    { superuser: true });
+                var buf = "";
+                this.monitor_channel.stream(output => {
+                    var lines;
```

Comment:
It's really a good idea to train yourself to use `let` and `const` instead of `var` in arrow functions and variables inside non-function code blocks, due to the weird scoping of `var`.

It doesn't cause trouble here, as the function is reasonably small, but just saying. For example, this would be a trap when calling an arrow function in a loop, `.filter()` or similar.

Reply:
> var moves out of the arrow function into the scope of the surrounding function

This *is* the case, and the reason why `var` is broken, and a trap for blocks/arrow functions in loops or callbacks. See https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Statements/var . Hoisting makes this even worse..

Reply:
I don't think it is true.  It certainly isn't true for non-arrow functions, and I am pretty sure that arrow functions behave like non-arrow functions with regard to `var`.

```
function foo() {
  const a = () => {
    var x = 12;
  }
  a();
  console.log(x);
}

foo()
ReferenceError: x is not defined
```

The trap with `var` and hoisting is that variables declared in a block scope are moved out of that scope to the function scope, and redeclarations are not an error:

```
function bar() {
  var x = 12;
  if (true) {
    var x = 14;
  }
  console.log(x);
}

bar()
14
```